File: Java/207.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] ind = new int[numCourses];
        Map<Integer, List<Integer>> map=new HashMap<>();
        for (int[] p : prerequisites){
            map.computeIfAbsent(p[1], o->new ArrayList<>()).add(p[0]);
            ind[p[0]]++;
        }
        Queue<Integer> queue = new ArrayDeque<>();
        for (int i = 0; i < ind.length; i++){
            if (ind[i]==0){
                queue.offer(i);
            }
        }
        while(!queue.isEmpty()){
            numCourses--;
            for (int next : map.getOrDefault(queue.poll(), List.of())){
                if (--ind[next]==0){
                    queue.offer(next);
                }
            }
        }
        return numCourses==0;
    }
}

File: C++/207.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& A) {
        /* vector<int> mp[numCourses]; */
        vector<vector<int>> mp(numCourses);
        vector<int> indg(numCourses);
        for (auto v : A){
            mp[v[1]].push_back(v[0]);
            ++indg[v[0]];
        }
        queue<int> q;
        for (int i = 0; i < numCourses; ++i){
            if (indg[i] == 0){
                q.push(i);
            }
        }
        while(!q.empty()){
            --numCourses;
            int cur = q.front();
            q.pop();
            for (int next : mp[cur]){
                if (--indg[next] == 0){
                    q.push(next);
                }
            }
        }
        return numCourses == 0;
    }
};
